test/IRGen/unmanaged_objc_throw_func.swift

```diff
@@ -0,0 +1,70 @@
+    // CHECK-NEXT: store i64 1, i64* %._value, align 8
+    // CHECK-NEXT: %7 = call swiftcc %TSo7NSArrayC* @"$sSa10FoundationE19_bridgeToObjectiveCSo7NSArrayCyF"(%swift.bridge* %4, %swift.type* @"$sSiN")
+    // CHECK-NEXT: %8 = bitcast %TSo7NSArrayC* %7 to %TSo10CFArrayRefa*
+    // CHECK-NEXT: call void asm sideeffect "", "r"(%TSo10CFArrayRefa* %8)
+    // CHECK-NEXT: call void @swift_bridgeObjectRelease(%swift.bridge* %4) #2
+    // CHECK-NEXT: %9 = bitcast %TSo10CFArrayRefa* %8 to i8*
+    // CHECK-NEXT: call void @llvm.objc.release(i8* %9)
+    // CHECK-NEXT: ret %TSo10CFArrayRefa* %8
+    let arr = [1] as CFArray
+    return Unmanaged.passUnretained(arr) 
+  } 
+}
+
+// CHECK: %14 = call swiftcc %TSo10CFArrayRefa* @"$s25unmanaged_objc_throw_func9SR_9035_CC22returnUnmanagedCFArrays0G0VySo0H3RefaGyKF"(%T25unmanaged_objc_throw_func9SR_9035_CC* swiftself %4, %swift.error** noalias nocapture swifterror dereferenceable(8) %swifterror) #6
+// CHECK-NEXT: %15 = load %swift.error*, %swift.error** %swifterror, align 8
+// CHECK-NEXT: %16 = icmp ne %swift.error* %15, null
+// CHECK-NEXT: br i1 %16, label %20, label %17
+
+// CHECK: ; <label>:17:                                     ; preds = %entry
+// CHECK-NEXT: %18 = phi %TSo10CFArrayRefa* [ %14, %entry ]
+// CHECK-NEXT: %19 = ptrtoint %TSo10CFArrayRefa* %18 to i64
+// CHECK-NEXT: br label %37
+
+// CHECK: ; <label>:20:                                     ; preds = %entry
+// CHECK-NEXT: %21 = phi %swift.error* [ %15, %entry ]
+// CHECK-NEXT: store %swift.error* null, %swift.error** %swifterror, align 8
+// CHECK-NEXT: %22 = icmp eq i64 %5, 0
+// CHECK-NEXT: br i1 %22, label %35, label %23
+
+// CHECK: ; <label>:23:                                     ; preds = %20
+// CHECK-NEXT: %24 = inttoptr i64 %5 to i8*
+// CHECK-NEXT: br label %25
+
+// CHECK: ; <label>:25:                                     ; preds = %23
+// CHECK-NEXT: %26 = phi i8* [ %24, %23 ]
+// CHECK-NEXT: %27 = call swiftcc %TSo7NSErrorC* @"$s10Foundation22_convertErrorToNSErrorySo0E0Cs0C0_pF"(%swift.error* %21) #6
+// CHECK: call swiftcc void @"$sSA7pointeexvs"(%swift.opaque* noalias nocapture %31, i8* %26, %swift.type* %33) #6
+// CHECK-NEXT: %34 = bitcast %TSo7NSErrorCSg* %3 to i8*
+// CHECK: call void @swift_errorRelease(%swift.error* %21) #2
+// CHECK-NEXT: br label %36
+
+// CHECK: ; <label>:35:                                     ; preds = %20
+// CHECK-NEXT: call void @swift_errorRelease(%swift.error* %21) #2
+// CHECK-NEXT: br label %36
+
+// CHECK: ; <label>:36:                                     ; preds = %25, %35
+// CHECK-NEXT: br label %37
+
+// CHECK: ; <label>:37:                                     ; preds = %17, %36
+// CHECK-NEXT: %38 = phi i64 [ 0, %36 ], [ %19, %17 ]
+// CHECK-NEXT: %39 = bitcast %T25unmanaged_objc_throw_func9SR_9035_CC* %4 to i8*
+// CHECK-NEXT: call void @llvm.objc.release(i8* %39)
+// CHECK-NEXT: %40 = inttoptr i64 %38 to %struct.__CFArray**
+// CHECK-NEXT: ret %struct.__CFArray** %40
```

Comment:
…and here's why we were testing it: this doesn't look like the right type. I think the *values* are all correct, but that type should have one fewer `*`, don't you think?

(Although maybe that's an existing bug. Worth checking against the behavior for plain `CFArray`.)

Reply:
Hmm if I return `arr` instead of `Unmanaged.passUnretained(arr)` then we get

```
%43 = inttoptr i64 %42 to %struct.__CFArray*
ret %struct.__CFArray* %43
```

Reply:
:-/ I guess we have a bug, then. How does the SIL look? Do we have extra indirection there?

Reply:
For `Unmanaged<CFArray>`, this is what the last couple of instructions look:

```asm
%14 = unchecked_ref_cast %13 : $NSArray to $CFArray // user: %15
%15 = unchecked_ref_cast %14 : $CFArray to $CFArray // users: %18, %16
debug_value %15 : $CFArray, let, name "arr"     // id: %16
release_value %6 : $Array<Int>                  // id: %17
%18 = ref_to_unmanaged %15 : $CFArray to $@sil_unmanaged CFArray // user: %19
%19 = struct $Unmanaged<CFArray> (%18 : $@sil_unmanaged CFArray) // user: %20
return %19 : $Unmanaged<CFArray>                // id: %20
```

For just `CFArray`:

```asm
%14 = unchecked_ref_cast %13 : $NSArray to $CFArray // user: %15
%15 = unchecked_ref_cast %14 : $CFArray to $CFArray // users: %18, %16
debug_value %15 : $CFArray, let, name "arr"     // id: %16
release_value %6 : $Array<Int>                  // id: %17
return %15 : $CFArray
```

Reply:
Ah, as before it's not this function that matters, but the one that bridges to ObjC.

Reply:
For `Unmanaged<CFArray>`, the thunk ends with
 
```asm
return %25 : $Optional<Unmanaged<CFArray>>
```

For `CFArray`

```asm
return %25 : $Optional<CFArray>
```

Reply:
Hmph. Both of those seem reasonable, so we must have a bug in IRGen. Drat! That means we can't just take this change.

Reply:
Oh! It seems to compile and run correctly however (at least in Swift). Anyway, I could file a JIRA for this if you want?

Reply:
I think it's still going to block *this* JIRA. It may not be causing problems now, but it could in the future (or possibly in a slightly more complicated test case).

You could test it a little harder in Swift by using `dynamic`, which will force the compiler to use `objc_msgSend` even for a Swift-defined method.

Reply:
Hmm. It works even after I add `dynamic` to the method. What would you suggest the best thing to do right now would be? Do we let this stay blocked until someone (maybe John?) takes a look at it or do we write more tests to ensure it behaves correctly?

Reply:
At this point I think someone has to look into why it's behaving incorrectly, _or_ show that it's not a regression.

You know, one thing we haven't tried is seeing what a normal `Optional<Unmanaged<CFArray>>` return looks like.

Reply:
IRGen:

```asm
%9 = ptrtoint %TSo10CFArrayRefa* %8 to i64
%10 = bitcast %TSo10CFArrayRefa* %8 to i8*
call void @llvm.objc.release(i8* %10)
ret i64 %9
```

SIL:

```
%18 = ref_to_unmanaged %15 : $CFArray to $@sil_unmanaged CFArray // user: %19
%19 = struct $Unmanaged<CFArray> (%18 : $@sil_unmanaged CFArray) // user: %20
%20 = enum $Optional<Unmanaged<CFArray>>, #Optional.some!enumelt.1, %19 : $Unmanaged<CFArray> // user: %22
strong_release %15 : $CFArray                   // id: %21
return %20 : $Optional<Unmanaged<CFArray>>
```

Reply:
\*shakes head\* Then yeah, something's going wrong specifically in the throwing case.

Reply:
That was without using `@objc`. If I use `@objc`, but remove the `throws`, then I get the same IR/SIL:

```
%13 = inttoptr i64 %11 to %struct.__CFArray**
ret %struct.__CFArray** %13
```

So it seems like the we get the same code with or without the `throws`. But if I remove `@objc`, then we get the code I mentioned in my previous comment.

Reply:
Bizarre, but okay, I guess we can treat it as a separate bug. (And this is a better reproducer for that bug.) Thanks for looking into it!